### ISSUE TITLE
Sidecar resource with workloadSelector only match Services

### ISSUE THREAD

**Issue by robertpanzer** [istio/istio#11818]
**Describe the bug**
When registering a Sidecar resource to restrict what a certain Pod can see, the restrictions are not applied if the spec contains a `workloadSelector`.

**Expected behavior**
When a Sidecar resource has a `workloadSelector`, it should be used to select pods for which the restrictions should be applied.

**Steps to reproduce the bug**
Take an empty Kubernetes cluster with Istio 1.1.0-snapshot.6.
Then apply this minimal configuration of one consumer – testcurl – and 2 provider services – testserver and testserver2 – in the namespaces `consumer` and `provider`, respectively:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  labels:
    istio-injection: enabled
  name: consumer

---
apiVersion: v1
kind: Namespace
metadata:
  labels:
    istio-injection: enabled
  name: provider

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: testcurl
  namespace: consumer
  labels:
    app: testcurl
spec:
  selector:
    matchLabels:
      app: testcurl
  replicas: 1
  template:
    metadata:
      labels:
        app: testcurl
    spec:
      containers:
        - name: testcurl
          image: tutum/curl
          command:
            - sleep
            - "1000000"

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: testserver
  namespace: provider
spec:
  selector:
    matchLabels:
      app: testserver
  replicas: 1
  template:
    metadata:
      labels:
        app: testserver
    spec:
      containers:
        - name: testserver
          image: kennethreitz/httpbin
          ports:
            - containerPort: 80
              name: http

---
apiVersion: v1
kind: Service
metadata:
  name: testserver
  namespace: provider
  labels:
    app: testserver
spec:
  selector:
    app: testserver
  ports:
    - name: http
      port: 8080
      protocol: TCP
      targetPort: http

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: testserver2
  namespace: provider
spec:
  selector:
    matchLabels:
      app: testserver2
  replicas: 1
  template:
    metadata:
      labels:
        app: testserver2
    spec:
      containers:
        - name: testserver
          image: kennethreitz/httpbin
          ports:
            - containerPort: 80
              name: http

---
apiVersion: v1
kind: Service
metadata:
  name: testserver2
  namespace: provider
  labels:
    app: testserver2

spec:
  selector:
    app: testserver2
  ports:
    - name: http
      port: 8080
      protocol: TCP
      targetPort: http
```

Open a shell in the testcurl container and test that a curl command to both providers return HTTP 200:
```
$ k exec -ti -n consumer testcurl-... -- curl -v testserver.provider:8080/get
$ k exec -ti -n consumer testcurl-... -- curl -v testserver2.provider:8080/get
```

Now create this sidecar resource:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Sidecar
metadata:
  name: testcurl
  namespace: consumer
spec:
  egress:
    - hosts:
        - "provider/testserver2.provider.svc.cluster.local"
```

After that testcurl is able to only connect to testserver2:
```
# This returns HTTP 404
$ k exec -ti -n consumer testcurl-... -- curl -v testserver.provider:8080/get 
# This returns HTTP 200
$ k exec -ti -n consumer testcurl-... -- curl -v testserver2.provider:8080/get
```

Now update the same sidecar resource so that it explicitly selects the consumer workload:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Sidecar
metadata:
  name: testcurl
  namespace: consumer
spec:
  workloadSelector:
    labels:
      app: testcurl
  egress:
    - hosts:
        - "provider/testserver2.provider.svc.cluster.local"
```
(I added the label `app=testcurl` to the Pod and Deployment just to make sure I didn't miss anything, I would expect though that the label on the Pod would be sufficient.)

Now try the same commands again:
```
# This should still return HTTP 404, but now it returns HTTP 200
$ k exec -ti -n consumer testcurl-... -- curl -v testserver.provider:8080/get 
# This returns HTTP 200
$ k exec -ti -n consumer testcurl-... -- curl -v testserver2.provider:8080/get
```

I expected the workloadSelector to still apply the restriction on egress traffic for the testcurl Pod, but as the result of the curl shows, it looks like traffic is not restricted anymore.

**Version**
Istio 1.1.0-snapshot.6 on Kubernetes 1.10.11 on Docker for Mac

**Installation**
```
helm install . --name istio --namespace istio-system \
  -f values-istio-auth-mcp.yaml  \
  --set global.mtls.enabled=true \
  --set global.tracing.enabled=true \
  --set pilot.traceSampling=100.0 \
  --set tracing.enabled=true \
  --set grafana.enabled=true
```

**Environment**
Local, Docker for Mac

**Cluster state**
{{ If you're running on Kubernetes, consider [following the
instructions](http://istio.io/help/bugs/#generating-a-cluster-state-archive)
to generate "istio-dump.tar.gz", then attach it here by dragging and dropping
the file onto this issue. }}

**Comment by rshriram** [comment]
could you delete the sidecar resource completely and retry with the new one (that has label selectors)? There was an issue earlier that the config updates to sidecar resources were not respected

**Comment by robertpanzer** [comment]
I tried directly creating the sidecar resource with the workload selector and it showed the same behavior, i.e. it allowed the call to `testserver.provider` even though.

Looking at the sources now it seems like this functionality is implemented in pilot/pkg/model/push_context.go:
https://github.com/istio/istio/blob/f62b4fa7d7ade9ab1c4a7d0307e649cc86d1434f/pilot/pkg/model/push_context.go#L455-L463

I'll try to dig deeper into this tomorrow.

I find this comment worrying though:
```
		// TODO: logic to merge multiple sidecar resources
		// Currently we assume that there will be only one sidecar config for a namespace.
```
Does that mean Sidecars are not ready to use yet?
Because as it's impossible to have multiple workloadSelectors, the only way to treat multiple workloads in one namespace differently  is to configure multiple Sidecars.

**Comment by hzxuzhonghu** [comment]
@rshriram @robertpanzer I think the root cause it `testCurl` not belong to any service so there is empty serviceInstances when calling 
`func (ps *PushContext) getSidecarScope(proxy *Proxy, proxyInstances []*ServiceInstance) *SidecarScope {
`

So that the workloadSelector does not take effect.

**Comment by robertpanzer** [comment]
Configuring a Service for my Deployment indeed fixed the problem.

But what is the reason for this? (Not in terms of implementation, but why)
My consumer deployment does not expose any ports. A Service though must have at least one Port, so I have to actually misconfigure it.

Also if this workloadSelector refers to a Service, I think the name 'workloadSelector' is unfortunate.
The policies vocabulary also uses the term `workload` in `source.workload.uid` and `destination.workload.uid`, and in this case it refers to the name of the Deployment.

**Comment by rshriram** [comment]
> Because as it's impossible to have multiple workloadSelectors, the only way to treat multiple workloads in one namespace differently is to configure multiple Sidecars.

that comment simply implies that if you have two or more sidecars with same workload Selectors or two or more sidecars with no workload selectors, there is no "merging" logic in place. You can definitely have multiple sidecars in a namespace as long as they all have different workload selectors.

**Comment by robertpanzer** [comment]
Yes, I figured that by now. I misread the comment.
I also tested that I can have two consumers in the same namespace with different egress clusters.

I still don't understand though why it is required to have a Service. 
Not every consumer is a provider at the same time.

**Comment by rshriram** [comment]
> I still don't understand though why it is required to have a Service.
> Not every consumer is a provider at the same time.

That is a valid point. I believe we had this restriction right from the start of the project. Though it has been removed recently, parts of the Pilot code (like the sidecar selection) still rely on this GetProxyServiceInstances() call. And as the name implies, it returns a non-empty array if and only if the proxy has ServiceInstances associated with it [i.e. pod is part of a service]. If it is not, it returns empty. 
This is a legacy API call that we haven't gotten rid of in Pilot.. ITs too late to make the change close to 1.1 release but we will try to get rid of it in the next patch release for 1.1

**Comment by munrodg** [comment]
@costinm setting to P1, do you think this needs to be a P0.

**Comment by hzxuzhonghu** [comment]
@rshriram I think we can fix this,  for k8s, sometimes we just need some instance info like labels/locality。 So i suggest defining a general structure independent of service. 

>This is a legacy API call that we haven't gotten rid of in Pilot.. ITs too late to make the change close to 1.1 release but we will try to get rid of it in the next patch release for 1.1

agreed

**Comment by rshriram** [comment]
ITs not that simple.. things like mTLS or rbac are defined on server side. and if a pod has no service instances, then these plugins can't be setup..

**Comment by howardjohn** [comment]
@rshriram can we just use `proxy.Metadata` here rather than get the labels of each serviceInstance? It seems like this makes more sense and would fix this issue. Is there any cases when we wouldn't want this behavior?

**Comment by duderino** [comment]
@howardjohn can you take this?

**Comment by robertpanzer** [comment]
I took a stab at this in #12620
No idea if this is sufficient or if I was missing important aspects as the PR looks so small.

**Comment by robertpanzer** [comment]
@howardjohn Taking the `proxy.Metadata` doesn't seem to work.
At least for me it's not updated on label changes to a pod.
It would need to create a new Envoy bootstrap config and restart it for that, right?

**Comment by robertpanzer** [comment]
The same problem also seems to exist for the EnvoyFilter selectors:
https://github.com/istio/istio/blob/190d4c1c59328e2aca2611f3056a03d45c583ca8/pilot/pkg/networking/core/v1alpha3/envoyfilter.go#L116-L126
ProxyInstances = ServiceInstances, which don't exist for consumer only services.

**Comment by howardjohn** [comment]
This seems really common, I noticed the same problem for locality based LB. If we have found 3 places, I am sure there are more too.

**Comment by hzxuzhonghu** [comment]
Currently one sidecar workload must have associated service. If not, then the pod healthcheck will fail.

**Comment by robertpanzer** [comment]
> Currently one sidecar workload must have associated service. If not, then the pod healthcheck will fail. 

Can you give a hint where that happens? Maybe I can do something about it.
This seems to be another bigger issue though.
Every application, also Pods without Services get a SidecarScope.

**Comment by hzxuzhonghu** [comment]
I mean this `func (p *Probe) checkInboundConfigured() error { `
in  `pilot/cmd/pilot-agent/status/ready/probe.go`

**Comment by robertpanzer** [comment]
I am not sure I understand the relation to selecting Sidecars for consumer applications given that there is always a Sidecar associated with Proxy.

I even have my PR deployed locally and configured a Sidecar to select my Pod, that is not exposed with a Service and I see this in the log of the istio-proxy container:
```
2019-03-21T08:01:05.842063Z     info    Envoy proxy is ready
```
The function that you mentioned returns no error if the application/pod does not expose any port, and indeed agent is started with an empty list:
```
    - --applicationPorts
    - ""
```

**Comment by hzxuzhonghu** [comment]
yep

**Comment by robertpanzer** [comment]
I created an alternative PR for this at https://github.com/istio/istio/pull/12666

**Comment by duderino** [comment]
@robertpanzer @howardjohn fix has been merged, can this be closed?

**Comment by howardjohn** [comment]
Yes, fix is merged in and will be released in 1.1.3 (expected around 4/15)
